urdf/src/urdf_plugin.cpp

```diff
@@ -0,0 +1,67 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2020, Willow Garage, Inc.
+*  All rights reserved.
+*
+*  Redistribution and use in source and binary forms, with or without
+*  modification, are permitted provided that the following conditions
+*  are met:
+*
+*   * Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*   * Redistributions in binary form must reproduce the above
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of the Willow Garage nor the names of its
+*     contributors may be used to endorse or promote products derived
+*     from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+#include <urdf_parser/urdf_parser.h>
+#include <urdf_parser_plugin/parser.h>
+
+#include <string>
+
+namespace urdf
+{
+class URDFXMLParser : public urdf::URDFParser
+{
+public:
+  URDFXMLParser() = default;
+
+  virtual ~URDFXMLParser() = default;
+
+  urdf::ModelInterfaceSharedPtr parse(const std::string & xml_string) override;
+
+  size_t might_handle(const std::string & data) override;
+};
+
+urdf::ModelInterfaceSharedPtr URDFXMLParser::parse(const std::string & xml_string)
+{
+  return urdf::parseURDF(xml_string);
+}
+
+size_t URDFXMLParser::might_handle(const std::string & data)
+{
+  // probably a urdf file if <robot is near the start of the document
+  return data.find("<robot");
```

Comment:
As I understand this, the closer `<robot` is to the beginning of the document, the higher the score it will get.

I think this is too simplistic.  In particular, adding a lot of comments at the beginning of an URDF file before the opening `<robot>` tag would cause the score to go way down, and I think that is pretty unintuitive for the user.

We're already depending on tinyxml2 in the upstream https://github.com/ros/urdf.  Can we add a dependency on that here as well, and then do a full parse on the XML document to check the root element?  It's obviously a lot more expensive, but I think it is also much more predictable.

Reply:
> As I understand this, the closer <robot is to the beginning of the document, the higher the score it will get.

nit: lower score, or higher confidence

> I think this is too simplistic. In particular, adding a lot of comments at the beginning of an URDF file before the opening <robot> tag would cause the score to go way down, and I think that is pretty unintuitive for the user.

It will make the score less confident, but the score doesn't need to be stable. It only needs to be more confident than the other parser plugins. I would have thought this approach is too simplistic too, but an even less sophisticated check that [picks the collada parser if `<COLLADA` appears anywhere in the document](https://github.com/ros/urdf/blob/e84c0ce88a37cf78ac456c7fae4b4e9bd7430069/urdf/src/model.cpp#L62) seems to have been fine for a long time.

>  Can we add a dependency on [tinyxml2] here as well, and then do a full parse on the XML document to check the root element? It's obviously a lot more expensive, but I think it is also much more predictable.

Implemented a check on the root element in cb348db to filter out the case where the string is a properly formed XML document that has a root tag other than `<robot>`

Reply:
> It will make the score less confident, but the score doesn't need to be stable. It only needs to be more confident than the other parser plugins. I would have thought this approach is too simplistic too, but an even less sophisticated check that [picks the collada parser if `<COLLADA` appears anywhere in the document](https://github.com/ros/urdf/blob/e84c0ce88a37cf78ac456c7fae4b4e9bd7430069/urdf/src/model.cpp#L62) seems to have been fine for a long time.

Agreed, but in the past, there was only URDF XML format and COLLADA.  With the addition of this new format (and maybe another one on the horizon), I think the chances of collision are going to become more likely.

> Implemented a check on the root element in [cb348db](https://github.com/ros2/urdf/commit/cb348dbd034d762b107cbeab862d05a2061c2c9d) to filter out the case where the string is a properly formed XML document that has a root tag other than `<robot>`

Thanks, that looks better (minus one nit which I'll point out in my second pass review).